workspace/notebook/py_sb_harmonised_nsip_meeting.json

```diff
@@ -0,0 +1,721 @@
+					"\n",
+					"            target_cols = target_df.columns\n",
+					"\n",
+					"            if \"NSIPProjectInfoInternalID\" not in target_cols:\n",
+					"                target_df = target_df.withColumn(\"NSIPProjectInfoInternalID\", lit(\"\"))\n",
+					"            \n",
+					"            # Get only active records for comparison\n",
+					"            target_active = target_df.filter(\"IsActive = 'Y'\")\n",
+					"            \n",
+					"            # Add hash for comparison\n",
+					"            target_active = target_active.withColumn(\n",
+					"                \"TargetHash\",\n",
+					"                md5(\n",
+					"                    concat_ws(\"|\",\n",
+					"                        coalesce(col(\"NSIPProjectInfoInternalID\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"caseId\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"caseReference\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"meetingAgenda\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"planningInspectorateRole\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"meetingDate\").cast(\"string\"), lit(\"\")),\n",
+					"                        coalesce(col(\"meetingType\").cast(\"string\"), lit(\"\"))\n",
+					"                    )\n",
+					"                )\n",
+					"            )\n",
+					"            \n",
+					"            # Get all historical records\n",
+					"            target_historical = target_df.filter(\"IsActive = 'N'\")\n",
+					"        else:\n",
+					"            # First run - no existing data\n",
+					"            target_active = None\n",
+					"            target_historical = None\n",
+					"            \n",
+					"    except Exception as e:\n",
+					"        error_message = f\"Error reading target data: {str(e)[:800]}\"\n",
+					"        end_exec_time = datetime.now()\n",
+					""
+				],
+				"execution_count": null
+			},
+			{
+				"cell_type": "markdown",
+				"metadata": {
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"##### Detect new and changed records with proper ValidTo dates for historical records"
+				]
+			},
+			{
+				"cell_type": "code",
+				"metadata": {
+					"jupyter": {
+						"source_hidden": false,
+						"outputs_hidden": false
+					},
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"if not error_message:\n",
+					"    try:\n",
+					"        current_timestamp = datetime.now()\n",
+					"        \n",
+					"        if target_active is not None:\n",
+					"            # Join source with active target to detect changes\n",
+					"            # 1. Ensure both source and target have the same natural grain\n",
+					"            key_cols = [\"meetingId\", \"meetingDate\"]\n",
```

Comment:
I think here the key_cols should only be meetingId and keep meetingDate inside the hash attributes (as you already do) but not in the join key.
Your own description at the top says : 
"meetingId as business key for SCD2"
Which implies that the join should be on meetingId only and the change detection should be done via the hash.

Shouldn't it be like this instead : 
key_cols = ["meetingId"]

comparison = source_df.alias("src").join(
    target_active.alias("tgt"),
    on=[col("src.meetingId") == col("tgt.meetingId")],
    how="full_outer"
)